### ISSUE TITLE
Creating derived rules from custom or existing rules

### ISSUE THREAD

**Issue by iankharris** [pyeve/cerberus#530]
Used Cerberus version / latest commit: …

- [x] I have the capacity to improve the docs when my problem is solved.
- [x] I have the capacity to submit a patch when a bug is identified.

- [ ] My question does not concern a practical use-case that I can't figure out
      to solve.

----

#### Use-case abstract

I would like to be able to extend existing rules from the rule_set_registry (custom or otherwise), with additional rules or constraints.  I've looked extensively at existing rules, and I can't figure out if this is yet possible -- and it should be.  The most obvious way to do this would be to use the `allof` rule to combine some rules, but 1) it nullifies the ability to normalize a document, and 2) rules do not qualify, only dictionaries (new schemas), which is a catch-22.

----

#### Bug report / Feature request

Let's say I have defined a rule:
```
rules_set_registry.extend({
    'some_rule': {
        'type': 'string',
    },
})
```

and in a later schema, I would like to extend (derive from) that rule; i.e., instead of 

```
schema = {
    'some_value': 'some_rule'
}
```

Let's say I want to add some additional rules to this.  So I would like to do something like

```
schema = {
    'some_value': {'allof': ['some_rule', {'required': True}]}
}
```

However, this throws an exception
```
    raise DocumentError(errors.DOCUMENT_FORMAT.format(document))
cerberus.validator.DocumentError: 'some_rule' is not a document, must be a dict
```

And prevents normalization of the document.  

So in other words, I'm looking for something equivalent to 
```
schema = {
    'some_value': dict(required=True, **rules_set_registry.get('some_rule'))
}
```

but hopefully less awkward.  (And I apologize if this is possible somehow, but if it is, it should at least be made clear in the documentation)

**Comment by funkyfuture** [comment]
yes, i have been thinking of 'stacking' *rule sets* and *schemas* (referenced or not), and i'm a little surprised that request comes so late.

i think the design would be simple: allow iterables that contain strings as references and/or mappings wherever these contained are allowed now and use an interpolation as constraint.

since some major refactorings are still due, please do not try to implement anything right now. but pull requests with tests that lay out the functionality would be a good base for discussions.

**Comment by iankharris** [comment]
I have since realized that this is semi-possible for schemas, by doing this:

```
schema = {
   'some_value': {'type': 'dict', 'schema': 'extended_schema', 'more': 'constraints'}
}
```

But frankly I have not tested this thoroughly yet.  If I find the time, I'd love to contribute to the project.  However it's not clear yet whether my team will be moving forwards with Cerberus or not -- I'm fighting for Cerberus currently :)

**Comment by Abhisek1994Roy** [comment]
Somewhat similar to what is mentioned here. Is there a possibility that one can use the inbuilt regex validation on integers or floats @funkyfuture ?

**Comment by funkyfuture** [comment]
i can't see how this is similar to this issue. and no, regular expressions have the capability to match strings. obviously you could cast any value to a string before checking it against a regex. i doubt that this would make sense though.

**Comment by funkyfuture** [comment]
@iankharris, could you elaborate what the `'more': 'constraints'` part is / does?

**Comment by iankharris** [comment]
@funkyfuture My apologies to everyone, I have been really busy the last couple months.  I should have been more helpful on this issue.  Anyway the `'more': 'constraints'` part just means any other constraints on your schema, for instance `defaults`.  In this way, you sort of 'extend' your ruleset using `extended_schema`

**Comment by funkyfuture** [comment]
we're not going to change Cerberus significantly, thus i'm closing this issue.
